src/algebra/order_operation_defs.lean

```diff
@@ -0,0 +1,291 @@
+Indeed, the usual ordered typeclasses arise from assuming the pair
+`[has_mul_le_mul_left M] [has_lt_of_mul_le_mul_left M]` on top of order/algebraic assumptions.
+
+A formal remark is that normally the relation that is an input to `covariant` is `(≤)`,
+while the relation that is an input to `contravariant` is `(<)`. This need not be the case in
+general, but seems to be the most common usage. In the opposite direction, the implication
+
+```lean
+[semigroup M] [partial_order M] [has_le_of_mul_le_mul M] → left_cancel_semigroup M
+```
+holds (note the `has_le_[...]` as opposed to the more common idioms `has_lt_[...]` and
+`has_mul_le_[...]`.
+-/
+-- TODO: convert `has_exists_mul_of_le`, `has_exists_add_of_le`?
+-- TODO: relationship with add_con
+-- include equivalence of `left_cancel_semigroup` with
+-- `semigroup partial_order contravariant_class α α (*) (≤)`?
+-- use ⇒, as per Eric's suggestion?
+section covariants_and_contravariants
+
+variables {M N : Type*} (μ : M → N → N) (r : N → N → Prop) (m : M) {a b c : N}
+
+/--  Let `M` and `N` be Types, with an action `μ : M → N → N` and a relation `r N → N → Prop`
+ on `N`.
+
+Informally, `covariant μ r` says that "the action `μ` preserves the relation `r`.
+
+More precisely, `covariant μ r` asserts that for all `m ∈ M` and all elements `n₁, n₂ ∈ N`, if the
+relation `r` holds for the pair `(n₁, n₂)`, then, the relation `r` also holds for the pair
+`(μ m n₁, μ m n₂)`, obtained from `(n₁, n₂)` by "acting upon it by `m`". -/
+def covariant     : Prop := ∀ (m) {n₁ n₂}, r n₁ n₂ → r (μ m n₁) (μ m n₂)
+
+/--  Let `M` and `N` be Types, with an action `μ : M → N → N` and a relation `r N → N → Prop`
+ on `N`.
+
+Informally, `contravariant μ r` says that "the action `μ` preserves the relation `r`.
+
+More precisely, `contravariant μ r` asserts that for all `m ∈ M` and all elements `n₁, n₂ ∈ N`,
+if the relation `r` holds for the pair `(μ m n₁, μ m n₂)`, obtained from `(n₁, n₂)` by
+"acting upon it by `m`", then, the relation `r` also holds for the pair `(n₁, n₂)`. -/
+def contravariant : Prop := ∀ (m) {n₁ n₂}, r (μ m n₁) (μ m n₂) → r n₁ n₂
+
+variables {M N}
+
+@[simp]
+lemma covariant_def :
+  covariant μ r ↔ ∀ (m) {n₁ n₂}, r n₁ n₂ → r (μ m n₁) (μ m n₂) :=
+iff.rfl
+
+@[simp]
+lemma contravariant_def :
+  contravariant μ r ↔ ∀ (m) {n₁ n₂}, r (μ m n₁) (μ m n₂) → r n₁ n₂ :=
+iff.rfl
+
+lemma covariant_iff_contravariant_le [linear_order N] :
+  covariant μ (≤) ↔ contravariant μ (<) :=
+⟨ λ h a b c bc, not_le.mp (λ k, not_le.mpr bc (h _ k)),
+  λ h a b c bc, not_lt.mp (λ k, not_lt.mpr bc (h _ k))⟩
+
+lemma covariant_iff_contravariant_lt [linear_order N] :
+  covariant μ (<) ↔ contravariant μ (≤) :=
+⟨ λ h a b c bc, not_lt.mp (λ k, not_lt.mpr bc (h _ k)),
+  λ h a b c bc, not_le.mp (λ k, not_le.mpr bc (h _ k))⟩
+
+
+lemma is_symm_op.swap_eq {α β} (op) [is_symm_op α β op] : flip op = op :=
+funext $ λ a, funext $ λ b, (is_symm_op.symm_op a b).symm
```

Comment:
This belongs somewhere way more basic; perhaps `function.basic` or even `logic.basic`.